nomad/jobs_endpoint.go

```diff
@@ -0,0 +1,297 @@
+	// we may get some extra efficiency, especially for non-contiguous job IDs.
+	if len(args.Jobs) > 0 {
+		nses := set.New[string](0)
+		for _, j := range args.Jobs {
+			nses.Insert(j.Namespace)
+		}
+		if nses.Size() == 1 {
+			namespace = nses.Slice()[0]
+		}
+	}
+
+	// Check for read-job permissions, since this endpoint includes alloc info
+	// and possibly a deployment ID, and those APIs require read-job.
+	aclObj, err := j.srv.ResolveACL(args)
+	if err != nil {
+		return err
+	}
+	if !aclObj.AllowNsOp(namespace, acl.NamespaceCapabilityReadJob) {
+		return structs.ErrPermissionDenied
+	}
+	allow := aclObj.AllowNsOpFunc(acl.NamespaceCapabilityReadJob)
+
+	// Compare between state run() unblocks to see if the RPC, as a whole,
+	// should unblock. i.e. if new jobs shift the page, or when jobs go away.
+	prevJobs := set.New[structs.NamespacedID](0)
+
+	sort := state.QueryOptionSort(args.QueryOptions)
+
+	// Setup the blocking query
+	opts := blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, state *state.StateStore) error {
+			var err error
+			var iter memdb.ResultIterator
+
+			// Get the namespaces the user is allowed to access.
+			allowableNamespaces, err := allowedNSes(aclObj, state, allow)
+			if errors.Is(err, structs.ErrPermissionDenied) {
+				// return empty jobs if token isn't authorized for any
+				// namespace, matching other endpoints
+				reply.Jobs = make([]structs.UIJob, 0)
+				return nil
+			} else if err != nil {
+				return err
+			}
+
+			if prefix := args.QueryOptions.Prefix; prefix != "" {
+				iter, err = state.JobsByIDPrefix(ws, namespace, prefix, sort)
+			} else if namespace != structs.AllNamespacesSentinel {
+				iter, err = state.JobsByNamespace(ws, namespace, sort)
+			} else {
+				iter, err = state.Jobs(ws, sort)
+			}
+			if err != nil {
+				return err
+			}
+
+			tokenizer := paginator.NewStructsTokenizer(
+				iter,
+				paginator.StructsTokenizerOptions{
+					WithNamespace: true,
+					WithID:        true,
+				},
+			)
+			filters := []paginator.Filter{
+				paginator.NamespaceFilter{
+					AllowableNamespaces: allowableNamespaces,
+				},
+				// skip child jobs; we'll look them up later, per parent.
+				paginator.GenericFilter{Allow: func(i interface{}) (bool, error) {
+					job := i.(*structs.Job)
+					return job.ParentID == "", nil
+				}},
+			}
+			// only provide specific jobs if requested.
+			if len(args.Jobs) > 0 {
+				// set per-page to avoid iterating the whole table
+				args.QueryOptions.PerPage = int32(len(args.Jobs))
+				// filter in the requested jobs
+				jobSet := set.From[structs.NamespacedID](args.Jobs)
+				filters = append(filters, paginator.GenericFilter{
+					Allow: func(i interface{}) (bool, error) {
+						job := i.(*structs.Job)
+						return jobSet.Contains(job.NamespacedID()), nil
+					},
+				})
+			}
+
+			jobs := make([]structs.UIJob, 0)
+			newJobs := set.New[structs.NamespacedID](0)
```

Comment:
Can we set these values with a starting point of `len(args.Jobs)` to avoid re-allocations?

Reply:
the caller could request invalid jobs, meaning the response will be smaller than the request.  but, I don't think that will be common, and it's still probably more efficient to avoid repeated mem allocs, as you say. :+1:

Reply:
turns out this is a bit funky.

I can do a little dance like this:

```go
i := 0
jobs := make([]structs.UIJob, len(args.Jobs))
newJobs := set.New[structs.NamespacedID](len(args.Jobs))

paginator..... {

	if len(args.Jobs) > 0 {
		jobs[i] = uiJob
		i++
	} else {
		jobs = append(jobs, uiJob)
	}
```

it feels strange, but it does work, _except_ when the caller asks for jobs that don't exist -- that ends up with zero-value `UIJob`s for each missing job (I could make them pointers, but missing jobs would show up in the results, as `nil`).  so I'd need to iterate over the results and exclude any empty ones, at which point I feel I've gone a little overboard in my optimization efforts.

maybe there's a trick here that I'm not aware of?

Reply:
I put that experiment in [3e0ab67](https://github.com/hashicorp/nomad/pull/20130/commits/3e0ab6718e5eda8275bc399f53c2a423393cc221), but was not a fan, so reverted.

let me know if you think it's worth it, or I should try it differently!